### PR TITLE
[ Besu ] add tls option for besu

### PR DESCRIPTION
**Pull request by lakshyakumar**
Signed-off-by: lakshyakumar <lakshya.kumar@accenture.com>

**Changelog**
- Add tls option for besu nodes.

 

**Reviewed by**
@jagpreetsinghsasan 

 

**Linked issue**
#939

### CODE REVIEW

**Comment by jagpreetsinghsasan** [comment]
Change the certificate paths in docs as well 
https://blockchain-automation-framework.readthedocs.io/en/latest/architectureref/certificates_path_list_besu.html

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml [hunk 1]
```diff
@@ -102,14 +102,40 @@
...
+    cd {{ ambassadortls }}
+    eval "openssl genrsa -out {{ node_name }}-besu-node.key 2048"
+    eval "openssl req -new -key {{ node_name }}-besu-node.key -out {{ node_name }}-besu-node.csr -subj '/CN={{ common_name }}'"
+    eval "openssl x509 -req -in {{ node_name }}-besu-node.csr -CA {{ rootca }}/rootca.pem -CAkey {{ rootca }}/rootca.key -CAcreateserial -out {{ node_name }}-besu-node.pem -days 500 -sha256"
+    eval "cat {{ node_name }}-besu-node.key {{ node_name }}-besu-node.pem > {{ node_name }}_certificates.pem"
+    eval "openssl pkcs12 -export -in {{ node_name }}_certificates.pem -out {{ node_name }}-besu-node.pkcs12 -password pass:somepass -name myAlias -noiter -nomaciter"
+    eval "echo "changeme" > {{ node_name }}-password"
+    eval "openssl x509 -in {{node_name}}-certchain.pem -noout -fingerprint -sha256 > {{node_name}}-certchain-sha256"
+    eval "export DIGEST=$(awk '{print $2}' {{node_name}}-certchain-sha256 | sed -n 's/Fingerprint=\(.*\)/\1/p')"
+    eval "echo "{{ common_name }} $DIGEST" > {{ node_name }}-knownServer"
```
Comment (sownak) [inline platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml:117]: Don't we have to also add the k8s internal address in the knownserver as well?

File: platforms/hyperledger-besu/charts/node_orion/templates/deployment.yaml [hunk 2]
```diff
@@ -197,6 +203,10 @@ spec:
             cat {{ .Values.node.mountPath }}/genesis/genesis.json.base64 | base64 -d > $BESU_GENESIS_DIR/genesis.json
             tr -d "\n\r" < /secrets/orion_key.pub > /secrets/nodekey.pub
             args="--rpc-http-enabled --rpc-http-api=ETH,NET,IBFT,DEBUG,ADMIN,WEB3,EEA,PRIV --rpc-ws-enabled --rpc-ws-api=ETH,NET,WEB3 --p2p-port {{ .Values.node.ports.p2p }} --rpc-http-port {{ .Values.node.ports.rpc }} --rpc-ws-port={{ .Values.node.ports.ws }}"
+            if [ {{ $.Values.orion.tls }} == "true" ]
+            then
+              tls_args="--privacy-tls-enabled --privacy-tls-keystore-file=/secrets/keystore.pkcs12 --privacy-tls-keystore-password-file=/secrets/keystore_password --privacy-tls-known-enclave-file=/secrets/known_server"
```
Comment (sownak) [inline platforms/hyperledger-besu/charts/node_orion/templates/deployment.yaml:208]: Where are the changes to orion.conf to accept tls connections from besu clients? Did you try deploying the private smart contract after these changes?

**Review by sownak — CHANGES_REQUESTED** [review]

File: platforms/hyperledger-besu/charts/node_orion/templates/deployment.yaml (hunk 2 above)
Comment (lakshyakumar) [reply platforms/hyperledger-besu/charts/node_orion/templates/deployment.yaml:208]: In the docs, the orion server configuration link is redirecting to the besu tls page, which is there in our code. I tried to deply the smart contract but it is still failing, looking into it.

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml (hunk 1 above)
Comment (lakshyakumar) [reply platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml]: added the hostname and the port in the knownserver file and  hostname in deployment for node_orion

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml [hunk 3]
```diff
@@ -102,14 +102,40 @@
...
+    eval "openssl req -new -key {{ node_name }}-besu-node.key -out {{ node_name }}-besu-node.csr -subj '/CN={{ common_name }}'"
+    eval "openssl x509 -req -in {{ node_name }}-besu-node.csr -CA {{ rootca }}/rootca.pem -CAkey {{ rootca }}/rootca.key -CAcreateserial -out {{ node_name }}-besu-node.pem -days 500 -sha256"
+    eval "cat {{ node_name }}-besu-node.key {{ node_name }}-besu-node.pem > {{ node_name }}_certificates.pem"
+    eval "openssl pkcs12 -export -in {{ node_name }}_certificates.pem -out {{ node_name }}-besu-node.pkcs12 -password pass:somepass -name myAlias -noiter -nomaciter"
+    eval "echo "changeme" > {{ node_name }}-password"
+    eval "openssl x509 -in {{node_name}}-certchain.pem -noout -fingerprint -sha256 > {{node_name}}-certchain-sha256"
+    eval "export DIGEST=$(awk '{print $2}' {{node_name}}-certchain-sha256 | sed -n 's/Fingerprint=\(.*\)/\1/p')"
+    eval "echo "{{ node_name }}:{{ node.tm_clientport.port | default('8888') }} $DIGEST" > {{ node_name }}-knownServer"
+  vars:
+    common_name: "{{ node_name }}.{{ organizationItem.external_url_suffix }}"
```
Comment (sownak) [inline platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml:119]: why is common_name removed from knownserver? I think both should be there

**Review by sownak — CHANGES_REQUESTED** [review]

File: platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml (hunk 3 above)
Comment (lakshyakumar) [reply platforms/hyperledger-besu/configuration/roles/create/certificates/ambassador/tasks/nested_main.yaml:119]: As per the besu docs the trusted enclave file is in format "<hostname>:<port> <hexstring>", so I changed the hostname as to match the hostname in deployment. Should it be changed to common name and the same common name is used in deployment?

**Comment by sownak** [comment]
Has this been tested by deploying the smartcontract on tls enabled network?

**Comment by lakshyakumar** [comment]
contract deployed
Smart contract is able to get deployed and contract address is generated on the besu network with tls.

**Review by sownak — APPROVED** [review]

**Pull request merged**